### ISSUE TITLE
C8y Log request plugin must publish the supported log types only c8y if bridge is up

### ISSUE THREAD

**Issue by PradeepKiruvale** [thin-edge/thin-edge.io#1184]
The `C8y log request plugin` should publish the supported `log types` only if the `c8y bridge` is created and up. Else publishing the supported `log type` message will not be reported to the c8y cloud.

**Describe the solution you'd like**

- The `c8y log request plugin` should subscribe to `tedge/health/mosquitto-c8y-bridge ` topic.
- If the bridge is `c8y bridge` is created and is up, then the `mqtt broker` will publish `1` on the `tedge/health/mosquitto-c8y-bridge` topic else it will publish `0`.
- When the `c8y log request plugin`  receives `1` on this topic, then it must publish the `supported log types` on `c8y/s/us` topic.
- Then the supported `log types` will be registered with the `c8y` cloud and it will help to request the required type of logs from the c8y cloud.

**Describe alternatives you've considered**

**Additional context**

**Comment by didier-wenzek** [comment]
I fully support this improvement. It would be good to the same thing for the configuration management plugin.

**Comment by rina23q** [comment]
Focus on only config management plugin and log file plugin. Later once we finish the implementation for the plugins, we should consider to apply the similar solution for mapper and agent.

**Comment by initard** [comment]
There is a new, unforeseen problem.

1. If you start `c8y_log_plugin`
2. connect to the cloud: `tedge connect c8y`
3. the `c8y_log_plugin` hangs

I believe the `c8y_log_plugin` hangs because mosquitto was restarted and the mqtt client object lost connection to it.

**Comment by PradeepKiruvale** [comment]
The `c8y_log_plugin` has to be started only after the `tedge connect c8y`. If the `c8y_log_plugin` running before `tedge connect c8y` it has to be restarted.

**Comment by rina23q** [comment]
While we were working on this issue, we encountered a bug, which seems the same as reported https://github.com/thin-edge/thin-edge.io/issues/1194.

The behaviour was observed in the version 0.7.2.

### How to reproduce?
[Precondition]
- All thin-edge packages are installed in the device.
- The device is already connected to c8y by `sudo tedge connect c8y`.
- The c8y_log_plugin process is up. Run the binary directly (execute `c8y_log_plugin --debug` in the shell). 

[Steps]
1. Publish a log request MQTT message. 
```
tedge mqtt pub 'c8y/s/ds' '522,DeviceSerial,logfileA,2013-06-22T17:03:14.000+02:00,2013-06-22T18:03:14.000+02:00,ERROR,1000'
```
2. Check the output of `c8y_log_plugin`. You can see that the plugin received the message by 1).
```
2022-07-15T13:08:08.564046292Z  INFO c8y_log_plugin: Log request received: 522,DeviceSerial,logfileA,2013-06-22T17:03:14.000+02:00,2013-06-22T18:03:14.000+02:00,ERROR,1000
```
3. Stop `mosquitto` service.
```
sudo systemctl stop mosquitto
```
4. Start `mosquitto` service.
```
sudo systemctl start mosquitto
```
5. Publish the same message as 1).
```
tedge mqtt pub 'c8y/s/ds' '522,DeviceSerial,logfileA,2013-06-22T17:03:14.000+02:00,2013-06-22T18:03:14.000+02:00,ERROR,1000'
```
6. Check the output of `c8y_log_plugin`. You can't see that the plugin received the new message by 5).

**Comment by rina23q** [comment]
Also, the same behaviour for `c8y_configuration_plugin`. Version 0.7.2.

### How to reproduce?
[Precondition]
- All thin-edge packages are installed in the device.
- The device is already connected to c8y by `sudo tedge connect c8y`.
- The c8y_configuration_plugin process is up. Run the binary directly (execute `c8y_configuration_plugin --debug` in the shell). 

[Steps]
1. Publish a configuration file change request MQTT message. 
```
tedge mqtt pub 'tedge/configuration_change/c8y-configuration-plugin' ''
```
2. Check the output of `c8y_configuration_plugin`. You can see that the plugin received the message by 1).
```
2022-07-15T13:14:31.673788063Z DEBUG c8y_configuration_plugin: Received Message { topic: Topic { name: "tedge/configuration_change/c8y-configuration-plugin" }, payload: [], qos: AtMostOnce, retain: false }
```
3. Stop `mosquitto` service.
```
sudo systemctl stop mosquitto
```
4. Start `mosquitto` service.
```
sudo systemctl start mosquitto
```
5. Publish the same message as 1).
```
tedge mqtt pub 'tedge/configuration_change/c8y-configuration-plugin' ''
```
6. Check the output of `c8y_configuration_plugin`. You can't see that the plugin received the new message by 5).

**Comment by rina23q** [comment]
By the way, `tedge_mapper c8y` can get recovered by restarting mosquitto. Even though the both plugins and mapper are using the same mqtt crate. So, I assume it's a bug on the plugins implementation.

**Comment by gligorisaev** [comment]
From all written here I cannot understand what is the actual status, missing any info how it was resolved

**Comment by initard** [comment]
From what I understand, the bug (@rina23q i don't know why this was labelled as a refactoring?) was addressed in #1240.

You will find in that PR two tests. The first test should pass (you should be able to test it). The second test does not pass. For convenience I pasted the tests here: 

## How to test

**Test 1: When bridge is up**
Given that the DUT is connected to c8y
- `sudo tedge connect c8y`

When the c8y_log_plugin is started
- `sudo systemctl start c8y-log-plugin.service`

Then c8y supported log types are uploaded to c8y
- cloud-side

**Test 2: When bridge is down**
Given that the DUT is NOT connected to c8y
- `sudo tedge disconnect c8y`

And the c8y_log_plugin is started
- `sudo systemctl start c8y-log-plugin.service`

When DUT is connected to c8y
- `sudo tedge connect c8y`

Then c8y supported log types are uploaded to c8y
- cloud-side


@rina23q is this how we close this issue? Or is it even closed since you said the same is true for the configuration plugin? Shouldn't that be another issue?

**Comment by rina23q** [comment]
> the bug (@rina23q i don't know why this was labelled as a refactoring?) 
Bug makes more sense.

**Comment by rina23q** [comment]
From my perspective, this issue is done. The change is not visible by end-user. The first test is sufficient to check this issue.

>Test 1: When bridge is up
Given that the DUT is connected to c8y
...

We need to create a new bug issue based on what I wrote [above](https://github.com/thin-edge/thin-edge.io/issues/1184#issuecomment-1185533832) and Test 2 by Alex.

**Comment by rina23q** [comment]
The PR #1240 changed the code based on our initial scope, presented by @PradeepKiruvale as [the first comment](https://github.com/thin-edge/thin-edge.io/issues/1184#issue-1264335856).

During the review of #1240, we found another bug as described [here](https://github.com/thin-edge/thin-edge.io/issues/1184#issuecomment-1185533832) and [here](https://github.com/thin-edge/thin-edge.io/issues/1184#issuecomment-1185537298). However, it is out of the initial scope of this issue, we will create a new issue.

The initial scope of this issue can be tested by "Test 1" proposed by [here](https://github.com/thin-edge/thin-edge.io/issues/1184#issuecomment-1191338668). "Test 2" is related to the bug that we will create newly.

**Comment by rina23q** [comment]
> During the review of https://github.com/thin-edge/thin-edge.io/pull/1240, we found another bug as described https://github.com/thin-edge/thin-edge.io/issues/1184#issuecomment-1185533832 and https://github.com/thin-edge/thin-edge.io/issues/1184#issuecomment-1185537298. However, it is out of the initial scope of this issue, we will create a new issue.

Created new issues, #1286 and #1287.
